File: factories.jl
# factories.jl
# Data types for a DSP factory and the groups of producers that make it up


#####################################
# RecipeCrafter Type Implementation #
#####################################

# When playing DSP, the number of facilities built obviously must be an Int
# However, we live in theory land where everything is perfectly "at ratio"
# To translate this fantasy number back into the DSP world, use ceil(howmany)
mutable struct RecipeCrafter
	recipe::Recipe
	facility::Facility
	howmany::Rational{Int}  # Int in DSP terms
	
	# TODO: proliferation info for recipe inputs
	#::Proliferator
	#::ProliferationMode
	
	rates::Dict{Product, Rate}
	
	function RecipeCrafter(recipe::Recipe, facility::Facility, howmany::Real)
		howmany > zero(howmany) ||
			error("RecipeProducer should not have howmany <= 0")
		new(recipe, facility, convert(Rational{Int}, howmany),
			Dict{Product, Rate}())
	end
end

# This constructor takes a production rate and computes the value of howmany
function RecipeCrafter(recipe::Recipe, facility::Facility,
	goal::ProductQuantity{Rate})
	
	p, rate = product(goal), quantity(goal)
	rc = RecipeCrafter(recipe, facility, 1)  # Temp value of howmany == 1
	
	haskey(rates(rc), p) && rates(rc)[p] > zero(Rate) ||
		error("given recipe does not produce goal product")
	
	rateratio = rate / rates(rc)[p]
	
	rc.howmany *= rateratio
	
	# TODO: consider proliferation of the recipe crafter
	
	_computerates!(rc)
	rc
end

recipe(rc::RecipeCrafter) = rc.recipe
facility(rc::RecipeCrafter) = rc.facility
howmany(rc::RecipeCrafter) = rc.howmany

# Note: this function needs to be careful about not overwriting Dict entries
# For example, X-Ray Cracking has Hydrogen inputs *and* outputs to account for
# Use get!() here so the output rate will not be overwritten by the input rate
function _computerates!(rc::RecipeCrafter)
	
	empty!(rc.rates)
	
	for output in outputs(recipe(rc))
		reciperate = quantity(output) / period(recipe(rc))
		totalrate = reciperate * speed(facility(rc)) * howmany(rc)
		rc.rates[product(output)] =
			get!(rc.rates, product(output), zero(Rate)) + totalrate
	end
	
	for input in inputs(recipe(rc))
		reciperate = quantity(input) / period(recipe(rc))
		totalrate = reciperate * speed(facility(rc)) * howmany(rc)
		rc.rates[product(input)] =
			get!(rc.rates, product(input), zero(Rate)) - totalrate
	end
	
	# TODO: proliferation would affect all these recipe rates
	# *and* it would consume proliferator product at a certain rate
	
	rc.rates
end

rates(rc::RecipeCrafter) = (isempty(rc.rates) && _computerates!(rc); rc.rates)

function Base.show(io::IO, rc::RecipeCrafter)
	print(io, "Group of ")
	print(io, prettystring(howmany(rc)))
	print(io, " ")
	print(io, name(facility(rc)))
	print(io, " crafting ")
	print(io, name(recipe(rc)))
end


###############################
# Factory Type Implementation #
###############################

struct Factory
	goal::ProductQuantity{Rate}
	crafters::Vector{RecipeCrafter}
	
	# TODO: output proflieration info for primary product
	#::Proliferator
	
	rates::Dict{Product, Rate}
	ignoredrates::Dict{Product, Bool}  # if true, ignore negative rates
	
	function Factory(goal::ProductQuantity{Rate})
		new(goal, Vector{RecipeCrafter}(), Dict{Product, Rate}(),
			Dict{Product, Bool}())
	end
end

goal(f::Factory) = f.goal
crafters(f::Factory) = f.crafters

function _computerates!(f::Factory)
	
	empty!(f.rates)
	
	f.rates[product(goal(f))] = -quantity(goal(f))
	mergewith!(+, f.rates, (rates(crafter) for crafter in crafters(f))...)
	
	f.rates
end

rates(f::Factory) = (isempty(f.rates) && _computerates!(f); f.rates)

function isignored(f::Factory, p::Product)
	haskey(rates(f), p) || error("product must exist to check if ignored")
	get!(f.ignoredrates, p, false)
end

# Negative rates are those which still need production added via crafters
# Ignored products have no recipe or they were shot down by the user
function negativerates(f::Factory)
	filter(rates(f)) do (product, productrate)
		!isignored(f, product) && productrate < zero(Rate)
	end
end

# Sometimes we need to know which crafters are producing a given product
function findcrafters(f::Factory, p::Product)
	filter(f.crafters) do crafter
		haskey(rates(crafter), p) && rates(crafter)[p] > zero(Rate)
	end
end

function Base.show(io::IO, f::Factory)
	
	INDENT = " " ^ 2
	
	print(io, "Factory:\n", INDENT, "Goal: ", goal(f))
	
	print(io, '\n', INDENT, "Crafters (", length(crafters(f)), "):")
	for crafter in crafters(f)
		print(io, '\n', INDENT ^ 2, crafter)
	end
	isempty(crafters(f)) && print(io, '\n', INDENT ^ 2, "(none)")
	
	inputs = filter(((product, rate),) -> rate < zero(Rate), rates(f))
	print(io, '\n', INDENT, "Inputs Required (", length(inputs), "):")
	for (product, productrate) in inputs
		print(io, '\n', INDENT ^ 2, ProductQuantity(-productrate, product))
	end
	isempty(inputs) && print(io, '\n', INDENT ^ 2, "(none)")
	
	byproducts = filter(((product, rate),) -> rate > zero(Rate), rates(f))
	print(io, '\n', INDENT, "Byproducts (", length(byproducts), "):")
	for (product, productrate) in byproducts
		print(io, '\n', INDENT ^ 2, ProductQuantity(productrate, product))
	end
	isempty(byproducts) && print(io, '\n', INDENT ^ 2, "(none)")
end


##################################
# Factory Manipulation Functions #
##################################

function setignored!(f::Factory, p::Product)
	haskey(rates(f), p) || error("product must exist to be set as ignored")
	f.ignoredrates[p] = true
	f
end

function setallignored!(f::Factory)
	for product in keys(rates(f))
		f.ignoredrates[product] = true
	end
	f
end

function connectcrafter!(f::Factory, rc::RecipeCrafter)
	rc in f.crafters && error("cannot connect same crafter to a factory twice")
	push!(f.crafters, rc)
	_computerates!(f)
	f
end

function upgradecrafter!(f::Factory, rc::RecipeCrafter,
	productionincrease::ProductQuantity{Rate})
	
	target = product(productionincrease)
	Δrate = quantity(productionincrease)
	
	rc in f.crafters || error("crafter must be factory-connected to upgrade")
	haskey(rates(rc), target) && rates(rc)[target] > zero(Rate) ||
		error("upgrading crafter does not produce target product")
	rates(rc)[target] + Δrate > zero(Rate) ||
		error("final upgraded crafter rate must be positive")
	
	upgraderatio = (rates(rc)[target] + Δrate) / rates(rc)[target]
	
	rc.howmany *= upgraderatio
	
	# TODO: consider how the proliferation of the crafter is to be upgraded
	
	_computerates!(rc)
	_computerates!(f)
	
	f
end

File: project_rustio/src/factories.rs
// factories.rs
// Data types for a DSP factory and the groups of producers that make it up

use std::collections::HashMap;
use std::fmt;

use num::rational::Rational64;
use num::traits::Zero;

use crate::facilities::Facility;
use crate::products::{Product, ProductQuantity};
use crate::rationalutilities::prettystring;
use crate::rates::Rate;
use crate::recipes::Recipe;


///////////////////////////////////////
// RecipeCrafter Type Implementation //
///////////////////////////////////////

// When playing DSP, the number of facilities built obviously must be an Int
// However, we live in theory land where everything is perfectly "at ratio"
// To translate this fantasy number back into the DSP world, use ceil(howmany)
#[derive(Debug, PartialEq, Eq)]
pub struct RecipeCrafter<'a> {
	recipe: Recipe,
	facility: Facility,
	howmany: Rational64,  // Int in DSP terms
	
	// TODO: proliferation info for recipe inputs
	// ::Proliferator
	// ::ProliferationMode
	
	rates: HashMap<&'a Product, Rate>,
}	

impl<'a> RecipeCrafter<'a> {
	pub fn new(recipe: Recipe, facility: Facility, howmany: Rational64) -> Self {
		if howmany <= Rational64::zero() {
			panic!("recipe producer should not have howmany <= 0");
		}
		Self {
			recipe,
			facility,
			howmany,
			rates: HashMap::new(),
		}
	}

	// This constructor takes a production rate and computes the value of howmany
	pub fn with_goal(recipe: Recipe, facility: Facility, goal: &ProductQuantity<Rate>) -> Self {
	
		let (p, rate) = (goal.product(), goal.quantity());
		let howmany = Rational64::new(1, 1);  // Temp value of 1
		let mut rc = Self::new(recipe, facility, howmany);
	
		match rc.rates().get(p) {
			Some(&r) if r > Rate::zero() => {
				let rateratio = rate / r;
				rc.howmany *= rateratio;
			},
			_ => panic!("given recipe does not produce goal product"),
		}
	
		// TODO: consider proliferation of the recipe crafter
	
		rc.computerates();
		rc
	}

	pub fn recipe(&self) -> &Recipe {
		&self.recipe
	}

	pub fn facility(&self) -> &Facility {
		&self.facility
	}
	
	pub fn howmany(&self) -> Rational64 {
		self.howmany
	}

	// Note: this function needs to be careful about not overwriting Dict entries
	// For example, X-Ray Cracking has Hydrogen inputs *and* outputs to account for
	// Use get!() here so the output rate will not be overwritten by the input rate
	fn computerates(&mut self) {
		self.rates.clear();
	
		for output in self.recipe().outputs() {
			let reciperate = output.quantity() / self.recipe().period();
			let totalrate = reciperate * self.facility().speed() * self.howmany();
			let accum = self.rates
			    .entry(output.product()).or_insert(Rate::zero());
			*accum += totalrate;
		}
	
		for input in self.recipe().inputs() {
			let reciperate = input.quantity() / self.recipe().period();
			let totalrate = reciperate * self.facility().speed() * self.howmany();
			let accum = self.rates
			    .entry(input.product()).or_insert(Rate::zero());
			*accum -= totalrate;
		}
	
		// TODO: proliferation would affect all these recipe rates
		// *and* it would consume proliferator product at a certain rate
	}

	pub fn rates(&mut self) -> &HashMap<&Product, Rate> {
		if self.rates.is_empty() {
			self.computerates();
		}
		&self.rates
	}
}

impl<'a> fmt::Display for RecipeCrafter<'a> {
	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
		write!(
			f,
			"Group of {} {} crafting {}",
			prettystring(self.howmany()),
			self.facility().name(),
			self.recipe().name()
		)
	}
}


/////////////////////////////////
// Factory Type Implementation //
/////////////////////////////////

#[derive(Debug)]
pub struct Factory<'a> {
	goal: ProductQuantity<Rate>,
	crafters: Vec<RecipeCrafter<'a>>,
	
	// TODO: output proflieration info for primary product
	// ::Proliferator
	
	rates: HashMap<&'a Product, Rate>,
	ignoredrates: HashMap<&'a Product, bool>,  // If true, ignore negative rates
}	

impl<'a> Factory<'a> {
	pub fn new(goal: ProductQuantity<Rate>) -> Self {
		Self {
			goal,
			crafters: Vec::new(),
			rates: HashMap::new(),
			ignoredrates: HashMap::new(),
		}
	}

	pub fn goal(&self) -> &ProductQuantity<Rate> {
		&self.goal
	}

	pub fn crafters(&self) -> &Vec<RecipeCrafter> {
		&self.crafters
	}

	fn computerates(&mut self) {
		self.rates.clear();
	
		self.rates.insert(
			self.goal().product(),
			-self.goal().quantity()
		);

		for crafter in self.crafters() {
			for (product, &rate) in crafter.rates() {
				let accum = self.rates
				    .entry(product).or_insert(Rate::zero());
				*accum += rate;
			}
		}
	}

	pub fn rates(&mut self) -> &HashMap<&Product, Rate> {
		if self.rates.is_empty() {
			self.computerates();
		}
		&self.rates
	}

	pub fn isignored(&mut self, p: &Product) -> bool {
		match self.rates().get(p) {
			None => panic!("product must exist to check if ignored"),
			_ => *self.ignoredrates.entry(p).or_insert(false),
		}
	}

	// Negative rates are those which still need production added via crafters
	// Ignored products have no recipe or they were shot down by the user
	pub fn negativerates(&self) -> Vec<(&Product, Rate)> {
		self.rates()
		    .iter()
			.filter(|(product, &productrate)| {
				!self.isignored(product) && productrate < Rate::zero()
			})
			.map(|(&product, &productrate)| {
				(product, productrate)
			})
			.collect()
	}

	// Sometimes we need to know which crafters are producing a given product
	pub fn findcrafters(&self, p: &Product) -> Vec<&RecipeCrafter> {
		self.crafters()
		    .iter()
		    .filter(|crafter| {
			    crafter.rates().contains_key(p) && *(crafter.rates().get(p).unwrap()) > Rate::zero()
		    })
		    .collect()
	}
}

impl<'a> fmt::Display for Factory<'a> {
	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
		let indent = " ".repeat(2);
		let mut io = String::new();

		io.push_str("Factory:\n");
		io.push_str(&indent);
		io.push_str("Goal: ");
		io.push_str(&self.goal().to_string());

		io.push_str("\n");
		io.push_str(&indent);
		io.push_str("Crafters (");
		io.push_str(&self.crafters().len().to_string());
		io.push_str("):");

		for crafter in self.crafters() {
			io.push_str("\n");
			io.push_str(&indent);
			io.push_str(&indent);
			io.push_str(&crafter.to_string())
		}

		if self.crafters().is_empty() {
			io.push_str("\n");
			io.push_str(&indent);
			io.push_str(&indent);
			io.push_str("(none)");
		}

		let inputs: Vec<(&Product, Rate)> = self
			.rates()
			.iter()
			.filter(|(product, &rate)| rate < Rate::zero())
			.map(|(&product, &rate)| (product, rate))
			.collect();

		io.push_str("\n");
		io.push_str(&indent);
		io.push_str("Inputs Required (");
		io.push_str(&inputs.len().to_string());
		io.push_str("):");

		for (product, productrate) in inputs.iter() {
			io.push_str("\n");
			io.push_str(&indent);
			io.push_str(&indent);
			let quantity = ProductQuantity::new(
				-(*productrate),
				*product.clone()
			);
			io.push_str(&quantity.to_string());
		}

		if inputs.is_empty() {
			io.push_str("\n");
			io.push_str(&indent);
			io.push_str(&indent);
			io.push_str("(none)");
		}

		let byproducts: Vec<(&Product, Rate)> = self
			.rates()
			.iter()
			.filter(|(product, &rate)| rate > Rate::zero())
			.map(|(product, rate)| (product.clone(), rate.clone()))
			.collect();

		io.push_str("\n");
		io.push_str(&indent);
		io.push_str("Byproducts (");
		io.push_str(&byproducts.len().to_string());
		io.push_str("):");

		for (product, productrate) in byproducts.iter() {
			io.push_str("\n");
			io.push_str(&indent);
			io.push_str(&indent);
			let quantity = ProductQuantity::new(
				*productrate,
				*product.clone()
			);
			io.push_str(&quantity.to_string());
		}

		if byproducts.is_empty() {
			io.push_str("\n");
			io.push_str(&indent);
			io.push_str(&indent);
			io.push_str("(none)");
		}

		write!(f, "{}", io)
	}
}


////////////////////////////////////
// Factory Manipulation Functions //
////////////////////////////////////

impl<'a> Factory<'a> {
	pub fn setignored(&mut self, p: &Product) {
		match self.rates().get_key_value(p) {
			None => panic!("product must exist to be set as ignored"),
			Some((&p, _)) => {
				self.ignoredrates.insert(p, true);
			},
		}
	}

	pub fn setallignored(&mut self) {
		for product in self.rates().keys() {
			self.ignoredrates.insert(product.clone(), true);
		}
	}

	pub fn connectcrafter(&mut self, rc: RecipeCrafter) {
		if self.crafters().contains(&rc) {
			panic!("cannot connect same crafter to a factory twice");
		}
		self.crafters.push(rc);
		self.computerates();
	}

	pub fn upgradecrafter(
		&mut self,
		mut rc: RecipeCrafter,
		productionincrease: ProductQuantity<Rate>
	) {
		
		let target = productionincrease.product();
		let delta_rate = productionincrease.quantity();
		
		if !self.crafters().contains(&rc) {
			panic!("crafter must be factory-connected to upgrade");
		}

		let rc_rates_target = rc.rates().get(target);
		if rc_rates_target.is_none() || *rc_rates_target.unwrap() <= Rate::zero() {
			panic!("upgrading crafter does not produce target product");
		}

		let rc_rates_target = rc_rates_target.unwrap();
		if *rc_rates_target + delta_rate <= Rate::zero() {
			panic!("final upgraded crafter rate must be positive");
		}
		
		let upgraderatio = (*rc_rates_target + delta_rate) / *rc_rates_target;
		
		rc.howmany *= upgraderatio;
		
		// TODO: consider how the proliferation of the crafter is to be upgraded
		
		rc.computerates();
		self.computerates();
	}
}
